File: classfiles/Domain/Dashboard/DashboardModels.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace MyWarehouse.Domain.Dashboard
{
    /// <summary>
    /// Dashboard configuration for a user
    /// </summary>
    public class DashboardConfiguration
    {
        [BsonId]
        [BsonRepresentation(BsonType.ObjectId)]
        public string Id { get; set; } = ObjectId.GenerateNewId().ToString();

        [BsonElement("UserId")]
        public int UserId { get; set; }

        [BsonElement("DashboardName")]
        public string DashboardName { get; set; } = string.Empty;

        [BsonElement("IsDefault")]
        public bool IsDefault { get; set; }

        [BsonElement("Layout")]
        public DashboardLayout Layout { get; set; } = new();

        [BsonElement("CreatedAt")]
        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

        [BsonElement("UpdatedAt")]
        public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
    }

    /// <summary>
    /// Dashboard layout configuration
    /// </summary>
    public class DashboardLayout
    {
        [BsonElement("columns")]
        public int Columns { get; set; } = 12;

        [BsonElement("rowHeight")]
        public int RowHeight { get; set; } = 80;

        [BsonElement("widgets")]
        public List<WidgetConfiguration> Widgets { get; set; } = new();
    }

    /// <summary>
    /// Widget configuration within a dashboard
    /// </summary>
    public class WidgetConfiguration
    {
        [BsonElement("widgetId")]
        public string WidgetId { get; set; } = string.Empty;

        [BsonElement("widgetType")]
        public string WidgetType { get; set; } = string.Empty;

        [BsonElement("position")]
        public WidgetPosition Position { get; set; } = new();

        [BsonElement("settings")]
        public Dictionary<string, object> Settings { get; set; } = new();
    }

    /// <summary>
    /// Widget position in the grid
    /// </summary>
    public class WidgetPosition
    {
        [BsonElement("x")]
        public int X { get; set; }

        [BsonElement("y")]
        public int Y { get; set; }

        [BsonElement("width")]
        public int Width { get; set; }

        [BsonElement("height")]
        public int Height { get; set; }
    }

    /// <summary>
    /// Widget definition in the library
    /// </summary>
    public class WidgetLibraryItem
    {
        [BsonId]
        [BsonRepresentation(BsonType.ObjectId)]
        public string Id { get; set; } = ObjectId.GenerateNewId().ToString();

        [BsonElement("WidgetId")]
        public string WidgetId { get; set; } = string.Empty;

        [BsonElement("WidgetType")]
        public string WidgetType { get; set; } = string.Empty;

        [BsonElement("Name")]
        public string Name { get; set; } = string.Empty;

        [BsonElement("Description")]
        public string Description { get; set; } = string.Empty;

        [BsonElement("Icon")]
        public string Icon { get; set; } = string.Empty;

        [BsonElement("Category")]
        public string Category { get; set; } = string.Empty;

        [BsonElement("DefaultSettings")]
        public Dictionary<string, object> DefaultSettings { get; set; } = new();

        [BsonElement("DefaultSize")]
        public WidgetSize DefaultSize { get; set; } = new();

        [BsonElement("MinSize")]
        public WidgetSize MinSize { get; set; } = new();

        [BsonElement("MaxSize")]
        public WidgetSize MaxSize { get; set; } = new();

        [BsonElement("RequiredPermissions")]
        public List<string> RequiredPermissions { get; set; } = new();

        [BsonElement("IsActive")]
        public bool IsActive { get; set; } = true;

        [BsonElement("CreatedAt")]
        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    }

    /// <summary>
    /// Widget size constraints
    /// </summary>
    public class WidgetSize
    {
        [BsonElement("width")]
        public int Width { get; set; }

        [BsonElement("height")]
        public int Height { get; set; }
    }

    /// <summary>
    /// Dashboard template
    /// </summary>
    public class DashboardTemplate
    {
        [BsonId]
        [BsonRepresentation(BsonType.ObjectId)]
        public string Id { get; set; } = ObjectId.GenerateNewId().ToString();

        [BsonElement("TemplateName")]
        public string TemplateName { get; set; } = string.Empty;

        [BsonElement("Description")]
        public string Description { get; set; } = string.Empty;

        [BsonElement("RoleId")]
        public int? RoleId { get; set; }

        [BsonElement("Layout")]
        public DashboardLayout Layout { get; set; } = new();

        [BsonElement("IsPublic")]
        public bool IsPublic { get; set; }

        [BsonElement("PreviewImage")]
        public string? PreviewImage { get; set; }

        [BsonElement("CreatedAt")]
        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    }
}

File: app/src/app/models/dashboard.models.ts
// Dashboard Models - matching backend DTOs

export interface WidgetPosition {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface WidgetConfiguration {
  widgetId: string;
  widgetType: string;
  position: WidgetPosition;
  settings: { [key: string]: any };
}

export interface DashboardLayout {
  columns: number;
  rowHeight: number;
  widgets: WidgetConfiguration[];
}

export interface DashboardConfiguration {
  id?: string;
  userId: number;
  dashboardName: string;
  isDefault: boolean;
  layout: DashboardLayout;
  createdAt?: Date;
  updatedAt?: Date;
}

export interface DashboardListItem {
  id: string;
  dashboardName: string;
  isDefault: boolean;
  widgetCount: number;
  createdAt: Date;
  updatedAt: Date;
}

export interface WidgetSize {
  width: number;
  height: number;
}

export interface WidgetLibraryItem {
  id?: string;
  widgetId: string;
  widgetType: string;
  name: string;
  description: string;
  icon: string;
  category: string;
  defaultSettings: { [key: string]: any };
  defaultSize: WidgetSize;
  minSize: WidgetSize;
  maxSize: WidgetSize;
  requiredPermissions: string[];
  isActive: boolean;
}

export interface DashboardTemplate {
  id?: string;
  templateName: string;
  description: string;
  roleId: number;
  layout: DashboardLayout;
  isPublic: boolean;
  previewImage?: string;
  createdAt?: Date;
}

// API Request/Response models
export interface SaveDashboardRequest {
  id?: string;
  userId: number;
  dashboardName: string;
  isDefault: boolean;
  layout: DashboardLayout;
}

export interface GetWidgetLibraryParams {
  category?: string;
  activeOnly?: boolean;
}

export interface GetDashboardTemplatesParams {
  roleId?: number;
  publicOnly?: boolean;
}

// ==========================================
// PHASE 4: REAL DATA MODELS
// ==========================================

/**
 * KPI Value Response from backend
 */
export interface KpiValueResponse {
  widgetId: string;
  value: number | string;
  showTrend: boolean;
  trendValue?: number;
  trendDirection?: 'up' | 'down';
  calculatedAt: Date;
}

/**
 * Activity Item Response
 */
export interface ActivityItemResponse {
  id: string;
  icon: string;
  iconColor: string;
  title: string;
  subtitle: string;
  timestamp: Date;
  metadata: string;
  activityType: string;
}

/**
 * Calendar Event Response
 */
export interface CalendarEventResponse {
  id: string;
  title: string;
  start: Date;
  end?: Date;
  color: string;
  type: string;
  description: string;
}

/**
 * Room Data for Room Planner
 */
export interface RoomData {
  id: string;
  roomId: number;
  roomNumber: string;
  roomName?: string;
  roomType: string;
  propertyId: number;
  propertyName: string;
  floor?: number;
  capacity?: number;
  status: string;
  amenities: string[];
  pricePerNight?: number;
  isActive: boolean;
}

/**
 * Booking with guest details response from backend
 */
export interface BookingWithGuestData {
  id: string;
  bookingId: string;
  roomNumber: string;
  propertyId: number;
  propertyName: string;
  checkInDate: Date | string;
  checkOutDate: Date | string;
  status: string;
  guestId: string;
  guestFirstName: string;
  guestLastName: string;
  guestEmail: string;
  guestPhoneNumber: string;
  guestNationality: string;
}

